tap_hubspot/__init__.py

```diff
@@ -699,7 +702,74 @@ def sync_deals(STATE, ctx):
     singer.write_state(STATE)
     return STATE
 
-#NB> no suitable bookmark is available: https://developers.hubspot.com/docs/methods/email/get_campaigns_by_id
+
+def gen_request_tickets(STATE, tap_stream_id, url, params, path, more_key):
+    """
+    Cursor-based API Pagination : Used in tickets stream implementation
+    """
+    with metrics.record_counter(tap_stream_id) as counter:
+        while True:
+            data = request(url, params).json()
+
+            if data.get(path) is None:
+                raise RuntimeError(
+                    "Unexpected API response: {} not in {}".format(path, data.keys()))
+
+            for row in data[path]:
+                counter.increment()
+                yield row
+
+            if not data.get(more_key):
+                break
+            params['after'] = data.get(more_key).get('next').get('after')
+
+    singer.write_state(STATE)
+
+
+def sync_tickets(STATE, ctx):
+    """
+    Function to sync `tickets` stream records
+    """
+    catalog = ctx.get_catalog_from_id(singer.get_currently_syncing(STATE))
+
+    stream_id = "tickets"
+    primary_key = "id"
+    bookmark_key = "updatedAt"
+
+    max_bk_value = bookmark_value = utils.strptime_with_tz(
+        get_start(STATE, stream_id, bookmark_key))
+    LOGGER.info("sync_tickets from %s", bookmark_value)
+
+    params = {'limit': 100,
+              'associations': 'contact,company,deals',
+              'properties': []}
+
+    schema = load_schema(stream_id)
+    singer.write_schema(stream_id, schema, [primary_key],
+                        [bookmark_key], catalog.get('stream_alias'))
+
+    url = get_url(stream_id)
+
+    with Transformer(UNIX_MILLISECONDS_INTEGER_DATETIME_PARSING):
+        for row in gen_request_tickets(STATE, stream_id, url, params, 'results', "paging"):
+
+            modified_time_org = row[bookmark_key]
+            modified_time = utils.strptime_with_tz(datetime.datetime.strptime(
+                modified_time_org, "%Y-%m-%dT%H:%M:%S.%fZ").strftime("%Y-%m-%dT%H:%M:%SZ"))
+
+            # Checking the bookmark value is present and the record is having
+            # grater than or equal to defined previous bookmark value
+            if modified_time and modified_time >= bookmark_value:
+                singer.write_record(stream_id, row, catalog.get(
+                    'stream_alias'), time_extracted=utils.now())
+                max_bk_value = modified_time
+
+    STATE = singer.write_bookmark(STATE, stream_id, bookmark_key, utils.strftime(max_bk_value))
+    singer.write_state(STATE)
+    return STATE
```

Comment:
I don't think this return statement is required

Reply:
`return STATE` is required. If you see line no. 1111, STREAM.sync(STATE, ctx) is expecting the return value `STATE` and if I remove return statement then it will return None and that will cause the issue. Hence, keeping it as it is.